src/edge_kernels/ScalarEdgeSolverAlg.C

```diff
@@ -0,0 +1,195 @@
+      const DblType viscosityR = dflux.get(nodeR, 0);
+
+      const DblType viscIp = 0.5 * (viscosityL + viscosityR);
+      const DblType diffIp = 0.5 * (viscosityL / densityL + viscosityR / densityR);
+
+      // Compute area vector related quantities and (U dot areaVec)
+      DblType axdx = 0.0;
+      DblType asq = 0.0;
+      DblType udotx = 0.0;
+      for (int d=0; d < ndim; ++d) {
+        const DblType dxj = coordinates.get(nodeR, d) - coordinates.get(nodeL, d);
+        asq += av[d] * av[d];
+        axdx += av[d] * dxj;
+        udotx += 0.5 * dxj * (vrtm.get(nodeR, d) + vrtm.get(nodeL, d));
+      }
+      const DblType inv_axdx = 1.0 / axdx;
+
+      // Compute extrapolated dq/dx
+      DblType dqL = 0.0;
+      DblType dqR = 0.0;
+      DblType nonOrth = 0.0;
+
+      for (int d=0; d < ndim; ++d) {
+        const DblType dxj = (coordinates.get(nodeR, d) - coordinates.get(nodeL, d));
+        dqL += 0.5 * dxj * dqdx.get(nodeL, d);
+        dqR += 0.5 * dxj * dqdx.get(nodeR, d);
+
+        const DblType kxj = av[d] - asq * inv_axdx * dxj;
+        nonOrth += -viscIp * kxj * 0.5 * (dqdx.get(nodeR, d) + dqdx.get(nodeL, d));
+      }
+
+      const DblType pecnum = stk::math::abs(udotx) / (diffIp + eps);
+      const DblType pecfac = pecFunc->execute(pecnum);
+      const DblType om_pecfac = 1.0 - pecfac;
+
+      DblType limitL = 1.0;
+      DblType limitR = 1.0;
+      if (useLimiter) {
+        const auto dq = qNp1R - qNp1L;
+        const auto dqML = 4.0 * dqL - dq;
+        const auto dqMR = 4.0 * dqR - dq;
+        limitL = van_leer(dqML, dq, eps);
+        limitR = van_leer(dqMR, dq, eps);
+      }
+
+      const DblType qIpL = qNp1L + dqL * hoUpwind * limitL;
+      const DblType qIpR = qNp1R - dqR * hoUpwind * limitR;
+
+      // Diffusive flux
+      const DblType lhsfac = -viscIp * asq * inv_axdx;
+      const DblType diffFlux = lhsfac * (qNp1R - qNp1L) + nonOrth;
+
+      // Left node
+      smdata.lhs(0, 0) = -lhsfac / relaxFac;
+      smdata.lhs(0, 1) = lhsfac;
+      smdata.rhs(0) = -diffFlux;
+      // Right node
+      smdata.lhs(1, 0) = lhsfac;
+      smdata.lhs(1, 1) = -lhsfac / relaxFac;
+      smdata.rhs(1) = diffFlux;
+
+      // Advective flux
+      const DblType qIp = 0.5 * (qNp1R + qNp1L); // 2nd order central term
+
+      // Upwinded term
+      const DblType qUpw = (mdot > 0)
+        ? (alphaUpw * qIpL + om_alphaUpw * qIp)
+        : (alphaUpw * qIpR + om_alphaUpw * qIp);
+
+      const DblType qHatL = (alpha * qIpL + om_alpha * qIp);
+      const DblType qHatR = (alpha * qIpR + om_alpha * qIp);
+      const DblType qCds = 0.5 * (qHatL + qHatR);
+
+      const DblType adv_flux = mdot * (pecfac * qUpw + om_pecfac * qCds);
+      smdata.rhs(0) -= adv_flux;
+      smdata.rhs(1) += adv_flux;
+
+      // Left node contribution; upwind terms
+      DblType alhsfac = 0.5 * (mdot + stk::math::abs(mdot))
+        * pecfac * alphaUpw + 0.5 * alpha * om_pecfac * mdot;
+      smdata.lhs(0, 0) += alhsfac / relaxFac;
```

Comment:
I wonder if it's worth storing alhsfac/relaxFac in a local since it gets used 3 times...

Reply:
@alanw0 Not sure I understand, `alhsfac` gets recomputed thrice before adding to `smdata.lhs`, and it does both with and without `relaxFac` for the L/R nodes per edge.

Reply:
@sayerhs yes I guess you're right. alhsfac/relaxFac gets used 4 times, but as you point out, alhsfac gets changed a couple times. alhsfac/relaxFac is only unchanged for the last 2, maybe it's not worth trying to save a divide or two.